README
rtl_hpsdr is an RTL to HPSDR software translation server.

It currently builds and runs on Linux. It identifies and uses up to
seven (theoretically eight) USB RTL2832U-based DVB-T dongles. The
dongles can be set up with an up converter or use RTL direct-mode
for HF receive. Or direct input to provide it's native > HF receive
range.

The program can be passed in a variety of command line options.
One of which is a frequency offset not only for up converter use
but also to allow a full range of frequency options to HPSDR programs
that are coded to only allow the real HPSDR radio's (i.e. Hermes)
frequency range which is from 10KHz to 55MHz.

The main purpose of this program is to provide a mechanism that
allows RTL Dongle owners the ability to use them on HPSDR specific
software programs.

One such program, cuSDR64 has the ability to control and display
up to 7 rcvr slices simultaneously. With rtl_hpsdr, if your host
has the horsepower, you can run 7 RTL Dongles to emulate the HPSDR
rcvr. cuSDR64 also can be built and run on Linux.

Since the real HPSDR (i.e. Hermes) rcvr can do up to eight rcvr
slices, there is a concept of 'COPY' rcvrs in this server. This
would allow one to use HPSDR programs that expected more rcvrs
than were attached. Currently if a program request more rcvrs
than are actually attached the rtl_hpsdr server will make copies
of the last 'real' rcvr. This alows one to only have one RTL dongle
attached and run PowerSDR mRX which may expect up to four rcvr
slices.

Refer to: http://openhpsdr.org/softwareinfo.php  for a list of
HPSDR supported applications.

I have tested this only on cuSDR64, cuSDR32, and PowerSDR mRX.

I was successful running 7 RTL dongles simultaneously on a
Quad core ARM Cortex A9 based mini-pc, model EKB311
running a version of Picuntu, http://ubuntu.g8.net/

Useful documentation:
http://openhpsdr.org/support/Ozy/USB_protocol_V1.48.pdf
http://svn.tapr.org/repos_sdr_hpsdr/trunk/Metis/Documentation/Metis-%20How%20it%20works_V1.30.pdf

Feel free to fix / add / modify this program.
I'd appreciate keeping me updated, n1gp@hotmail.com


1. Dependencies

- At least one (and up to seven) RTL2832U-based DVB-T dongles
- RTL-SDR from http://cgit.osmocom.org/cgit/rtl-sdr/


2. Installation

Get the latest code from: https://github.com/n1gp/rtl_hpsdr

The Makefile is setup to detect the CPU NEON or SSE2
SIMD capabilities. If none are found it will compile
without optimizations for the downsampling routines.

$ make

3. Usage

rtl_hpsdr can be started using command line options or with a
confuration file. See rtl_hpsdr.conf for an example.

$ ./rtl_hpsdr -c rtl_hpsdr.conf

For help

$ ./rtl_hpsdr -h

4. Known problems

- This server program needs to be run on a separate PC than
  the client program. Probably due to improper network settings
  in the reveal and discovery socket layers or perhaps its not
  possible to connect from the same client port to the same
  server (address and port). [ FIXED in version 1.3 ]

- Periodically the Alsa local sound skips and is caused by xruns.
  Possibly due to improper period and buffer settings.

- The down sampling routines have not been validated for accuracy
  and perhaps could be better optimized.

5. Credits and License

rtl_hpsdr is designed and written by Richard Koch N1GP,
and is licensed under the GNU General Public License.

I give thanks and credit to the http://openhpsdr.org community.
All the folks who made the rtl-sdr libraries and utilities.
The work of John Melton, G0ORX/N6LYT and others on
the ghpsdr3 open source projects which provided many examples
for communicating with HPSDR. Some alsa audio routines
I borrowed from Matthieu Sozeau and the XMMS project. Also
Olli Parviainen and his application SoundTouch where I got idea's
on optimizing the downsampling filter routines using SIMD
instructions, http://www.surina.net/soundtouch




ChangeLog
2014-01-26  Richard Koch  <n1gp@hotmail.com>

* 1.3,	Fixed issue with running this server and a client on the same PC

* 1.2,	Added support for CW SkimSrv: http://dxatlas.com/SkimServer/
	This uses Vasiliy K3IT's HermesIntf.dll which he modified
	for rtl_hpsdr so it can use up to 8 dongles to cover 8 bands.
	Get that here: http://sourceforge.net/projects/hermesintf/files/

2014-01-24  Richard Koch  <n1gp@hotmail.com>

* 1.1,	Added rcvr_order

2014-01-13  Richard Koch  <n1gp@hotmail.com>

* 1.0,	Initial release



coeff.c
/* FIR filters designed with
   http://t-filter.appspot.com

sampling frequencies are based on
2 * RTL sample rate for nyquist

*/

/*
sampling frequency: 3072000 Hz

* 0 Hz - 370000 Hz
  gain = 1
  desired ripple = 1 dB
  actual ripple = 0.21686021954788914 dB

* 800000 Hz - 1536000 Hz
  gain = 0
  desired attenuation = -40 dB
  actual attenuation = -50.79615251277994 dB
*/

#include "rtl_hpsdr.h"

// 768Khz
float coeff3072_768_H_16[COEFF3072_H_16_LENGTH] = {
	0.009273441360506243,
	0.016647116642970215,
	0.00017807828712343627,
	-0.04120881516829146,
	-0.054182354998635315,
	0.029186072429927865,
	0.19986560493298844,
	0.3445483364271308,
	0.3445483364271308,
	0.19986560493298844,
	0.029186072429927865,
	-0.054182354998635315,
	-0.04120881516829146,
	0.00017807828712343627,
	0.016647116642970215,
	0.009273441360506243
};

/*
sampling frequency: 1536000 Hz

* 0 Hz - 360000 Hz
  gain = 1
  desired ripple = 1 dB
  actual ripple = 1.339203288338704 dB

* 420000 Hz - 768000 Hz
  gain = 0
  desired attenuation = -40 dB
  actual attenuation = -34.99955193016408 dB
*/
// 384Khz
float coeff1536_384_H_32[COEFF1536_H_32_LENGTH] = {
	-0.018813375437660432,
	-0.006245528066330151,
	0.02225837116892513,
	0.02542479094783578,
	-0.005640794972019027,
	-0.012689256912544498,
	0.022189644598742682,
	0.02694074050135046,
	-0.02278035343087979,
	-0.031038461875351708,
	0.038216616239274616,
	0.050231714731015474,
	-0.06031941099781234,
	-0.08957130195006367,
	0.14839312870816138,
	0.45191266878865377,
	0.45191266878865377,
	0.14839312870816138,
	-0.08957130195006367,
	-0.06031941099781234,
	0.050231714731015474,
	0.038216616239274616,
	-0.031038461875351708,
	-0.02278035343087979,
	0.02694074050135046,
	0.022189644598742682,
	-0.012689256912544498,
	-0.005640794972019027,
	0.02542479094783578,
	0.02225837116892513,
	-0.006245528066330151,
	-0.018813375437660432
};

/*
sampling frequency: 1536000 Hz

* 0 Hz - 170000 Hz
  gain = 1
  desired ripple = 1 dB
  actual ripple = 1.3804853552655727 dB

* 230000 Hz - 768000 Hz
  gain = 0
  desired attenuation = -40 dB
  actual attenuation = -34.73692009329559 dB
*/
// 192Khz
float coeff1536_192_H_32[COEFF1536_H_32_LENGTH] = {
	-0.010042974482114305,
	-0.023419612995074016,
	-0.01704587811295064,
	-0.013595519567577328,
	0.004738092999209698,
	0.019741277216946216,
	0.026934187646518025,
	0.015691447421015893,
	-0.010353179647232053,
	-0.039415885401377665,
	-0.05142493084837827,
	-0.02988367769651434,
	0.02970163874460875,
	0.11419149328873458,
	0.1967846501935727,
	0.24774873746989148,
	0.24774873746989148,
	0.1967846501935727,
	0.11419149328873458,
	0.02970163874460875,
	-0.02988367769651434,
	-0.05142493084837827,
	-0.039415885401377665,
	-0.010353179647232053,
	0.015691447421015893,
	0.026934187646518025,
	0.019741277216946216,
	0.004738092999209698,
	-0.013595519567577328,
	-0.01704587811295064,
	-0.023419612995074016,
	-0.010042974482114305
};

/*
sampling frequency: 1536000 Hz

* 0 Hz - 70000 Hz
  gain = 1
  desired ripple = 1 dB
  actual ripple = 1.969461831682125 dB

* 120000 Hz - 768000 Hz
  gain = 0
  desired attenuation = -40 dB
  actual attenuation = -31.669450042094866 dB
*/
// 96Khz
float coeff1536_96_H_32[COEFF1536_H_32_LENGTH] = {
	-0.019744354306438096,
	-0.015506545711307839,
	-0.01925410128922757,
	-0.021493399566871418,
	-0.021345056921586145,
	-0.01806406451421989,
	-0.01115385429515886,
	-0.0004477980642554994,
	0.013807116545973127,
	0.03090920205193141,
	0.049747010844936115,
	0.06892032225707885,
	0.08686969322593223,
	0.10200900320293706,
	0.11295785369204779,
	0.11872097296819287,
	0.11872097296819287,
	0.11295785369204779,
	0.10200900320293706,
	0.08686969322593223,
	0.06892032225707885,
	0.049747010844936115,
	0.03090920205193141,
	0.013807116545973127,
	-0.0004477980642554994,
	-0.01115385429515886,
	-0.01806406451421989,
	-0.021345056921586145,
	-0.021493399566871418,
	-0.01925410128922757,
	-0.015506545711307839,
	-0.019744354306438096
};

/*
sampling frequency: 1536000 Hz

* 0 Hz - 40000 Hz
  gain = 1
  desired ripple = 5 dB
  actual ripple = 4.876115918200986 dB

* 90000 Hz - 768000 Hz
  gain = 0
  desired attenuation = -40 dB
  actual attenuation = -38.71256411909809 dB
*/
// 48Khz
float coeff1536_48_H_32[COEFF1536_H_32_LENGTH] = {
	0.0002638580125835885,
	0.009790810578691002,
	0.009414466063966724,
	0.013631551091975282,
	0.01848634664234968,
	0.02401139026418836,
	0.030078953329934023,
	0.03652064137017532,
	0.04312893292488267,
	0.04968081967326505,
	0.05591992373138862,
	0.06160415916164228,
	0.06647785250512578,
	0.07034509358622862,
	0.0730269248498467,
	0.07439089233305521,
	0.07439089233305521,
	0.0730269248498467,
	0.07034509358622862,
	0.06647785250512578,
	0.06160415916164228,
	0.05591992373138862,
	0.04968081967326505,
	0.04312893292488267,
	0.03652064137017532,
	0.030078953329934023,
	0.02401139026418836,
	0.01848634664234968,
	0.013631551091975282,
	0.009414466063966724,
	0.009790810578691002,
	0.0002638580125835885
};

/*
sampling frequency: 1536000 Hz

* 0 Hz - 368000 Hz
  gain = 1
  desired ripple = 1 dB
  actual ripple = 0.8363669011491076 dB

* 404000 Hz - 768000 Hz
  gain = 0
  desired attenuation = -40 dB
  actual attenuation = -39.078069987820086 dB
*/
// 384Khz
float coeff1536_384_H_64[COEFF1536_H_64_LENGTH] = {
	-0.006749871284054873,
	0.004480925740369876,
	0.01484913649951335,
	0.01141731022874683,
	-0.0021395270640331694,
	-0.004146112552536018,
	0.006108304018932493,
	0.006023128067669043,
	-0.006385590920724013,
	-0.006489590558851686,
	0.007672161073696956,
	0.0076536235677186395,
	-0.008966828328483275,
	-0.009014764701078176,
	0.010553692980521202,
	0.010722878179753304,
	-0.012481901630215978,
	-0.012862783455557508,
	0.014911904711552218,
	0.015629462382057684,
	-0.018103314264455436,
	-0.019358200988217264,
	0.02252835451759192,
	0.024712778310252446,
	-0.029154241879223,
	-0.03316745807801754,
	0.04039633773067006,
	0.04888789948915267,
	-0.06410538838571432,
	-0.0892335106077302,
	0.15017819477314162,
	0.44968698442469557,
	0.44968698442469557,
	0.15017819477314162,
	-0.0892335106077302,
	-0.06410538838571432,
	0.04888789948915267,
	0.04039633773067006,
	-0.03316745807801754,
	-0.029154241879223,
	0.024712778310252446,
	0.02252835451759192,
	-0.019358200988217264,
	-0.018103314264455436,
	0.015629462382057684,
	0.014911904711552218,
	-0.012862783455557508,
	-0.012481901630215978,
	0.010722878179753304,
	0.010553692980521202,
	-0.009014764701078176,
	-0.008966828328483275,
	0.0076536235677186395,
	0.007672161073696956,
	-0.006489590558851686,
	-0.006385590920724013,
	0.006023128067669043,
	0.006108304018932493,
	-0.004146112552536018,
	-0.0021395270640331694,
	0.01141731022874683,
	0.01484913649951335,
	0.004480925740369876,
	-0.006749871284054873
};

/*
sampling frequency: 1536000 Hz

* 0 Hz - 170000 Hz
  gain = 1
  desired ripple = 1 dB
  actual ripple = 0.9493604453903074 dB

* 202000 Hz - 768000 Hz
  gain = 0
  desired attenuation = -36 dB
  actual attenuation = -33.97931753861141 dB
*/
// 192Khz
float coeff1536_192_H_64[COEFF1536_H_64_LENGTH] = {
	-0.010861343163249607,
	-0.00018869317372753413,
	0.0038397866507845894,
	0.00874283484070055,
	0.01191201638889358,
	0.011282422852842505,
	0.006424834873306104,
	-0.0008014692537745406,
	-0.006952478489948671,
	-0.008683753210373287,
	-0.004679440396557242,
	0.0033281946583610575,
	0.011078123345028841,
	0.013841539912365117,
	0.00907043609384835,
	-0.001762715597699286,
	-0.013365798458316624,
	-0.019052763404783082,
	-0.014372799752652075,
	-0.00007240060352242824,
	0.017348049196545607,
	0.028206736320385815,
	0.024468683175278203,
	0.0046265889006657495,
	-0.02394839255527137,
	-0.0467951590408317,
	-0.04800971319194729,
	-0.017509854725669075,
	0.043342515614600334,
	0.12056900886575762,
	0.19180649750128276,
	0.23446612587689916,
	0.23446612587689916,
	0.19180649750128276,
	0.12056900886575762,
	0.043342515614600334,
	-0.017509854725669075,
	-0.04800971319194729,
	-0.0467951590408317,
	-0.02394839255527137,
	0.0046265889006657495,
	0.024468683175278203,
	0.028206736320385815,
	0.017348049196545607,
	-0.00007240060352242824,
	-0.014372799752652075,
	-0.019052763404783082,
	-0.013365798458316624,
	-0.001762715597699286,
	0.00907043609384835,
	0.013841539912365117,
	0.011078123345028841,
	0.0033281946583610575,
	-0.004679440396557242,
	-0.008683753210373287,
	-0.006952478489948671,
	-0.0008014692537745406,
	0.006424834873306104,
	0.011282422852842505,
	0.01191201638889358,
	0.00874283484070055,
	0.0038397866507845894,
	-0.00018869317372753413,
	-0.010861343163249607
};

/*
sampling frequency: 1536000 Hz

* 0 Hz - 76000 Hz
  gain = 1
  desired ripple = 1 dB
  actual ripple = 0.7698988848526747 dB

* 110000 Hz - 768000 Hz
  gain = 0
  desired attenuation = -36 dB
  actual attenuation = -35.796309630710496 dB
*/
// 96Khz
float coeff1536_96_H_64[COEFF1536_H_64_LENGTH] = {
	-0.01071252749483098,
	-0.005596767001554159,
	-0.006134010497622714,
	-0.005932358028180867,
	-0.004903426703490449,
	-0.002987594254311014,
	-0.00032198482516542,
	0.002838867548545009,
	0.0061196328186719365,
	0.009067056230273823,
	0.011180510006145467,
	0.012006372127994101,
	0.011211344812046414,
	0.00865134713119215,
	0.00440832129964661,
	-0.0011745304013948542,
	-0.007496596163064408,
	-0.013739689527000921,
	-0.018953804529181623,
	-0.02215844996076096,
	-0.02245960865939106,
	-0.019165008103936735,
	-0.011892104124015358,
	-0.0006430580031297027,
	0.014159357426518738,
	0.031684756028006196,
	0.050753304724008914,
	0.06994068895712066,
	0.08771620731707629,
	0.1025991616107944,
	0.11331223097868615,
	0.11891723869590523,
	0.11891723869590523,
	0.11331223097868615,
	0.1025991616107944,
	0.08771620731707629,
	0.06994068895712066,
	0.050753304724008914,
	0.031684756028006196,
	0.014159357426518738,
	-0.0006430580031297027,
	-0.011892104124015358,
	-0.019165008103936735,
	-0.02245960865939106,
	-0.02215844996076096,
	-0.018953804529181623,
	-0.013739689527000921,
	-0.007496596163064408,
	-0.0011745304013948542,
	0.00440832129964661,
	0.00865134713119215,
	0.011211344812046414,
	0.012006372127994101,
	0.011180510006145467,
	0.009067056230273823,
	0.0061196328186719365,
	0.002838867548545009,
	-0.00032198482516542,
	-0.002987594254311014,
	-0.004903426703490449,
	-0.005932358028180867,
	-0.006134010497622714,
	-0.005596767001554159,
	-0.01071252749483098
};

/*
sampling frequency: 1536000 Hz

* 0 Hz - 40000 Hz
  gain = 1
  desired ripple = 1 dB
  actual ripple = 1.829962937463533 dB

* 66000 Hz - 768000 Hz
  gain = 0
  desired attenuation = -36 dB
  actual attenuation = -28.30249878265792 dB
*/
// 48Khz
float coeff1536_48_H_64[COEFF1536_H_64_LENGTH] = {
	0.016204784472223264,
	-0.005952326048019109,
	-0.005980133480946519,
	-0.006550587340783938,
	-0.007502349558746014,
	-0.008664905618739272,
	-0.009858201134284412,
	-0.011058172934685848,
	-0.012052095453511394,
	-0.012771304376233453,
	-0.013107166855460603,
	-0.012962322150519364,
	-0.012244684255291452,
	-0.01091202402746656,
	-0.008896557288651442,
	-0.006183978870649281,
	-0.0027753248823173416,
	0.0013053700995675575,
	0.006010078044350416,
	0.011250745337822507,
	0.01693463711646011,
	0.02293456274088121,
	0.02910743703535114,
	0.03530091736205998,
	0.04135453695003249,
	0.04709815437714355,
	0.05237637339324536,
	0.05703683825323971,
	0.06094137268623868,
	0.06397734150508055,
	0.06605476099202433,
	0.0671076198746522,
	0.0671076198746522,
	0.06605476099202433,
	0.06397734150508055,
	0.06094137268623868,
	0.05703683825323971,
	0.05237637339324536,
	0.04709815437714355,
	0.04135453695003249,
	0.03530091736205998,
	0.02910743703535114,
	0.02293456274088121,
	0.01693463711646011,
	0.011250745337822507,
	0.006010078044350416,
	0.0013053700995675575,
	-0.0027753248823173416,
	-0.006183978870649281,
	-0.008896557288651442,
	-0.01091202402746656,
	-0.012244684255291452,
	-0.012962322150519364,
	-0.013107166855460603,
	-0.012771304376233453,
	-0.012052095453511394,
	-0.011058172934685848,
	-0.009858201134284412,
	-0.008664905618739272,
	-0.007502349558746014,
	-0.006550587340783938,
	-0.005980133480946519,
	-0.005952326048019109,
	0.016204784472223264
};


downsample.c
/*	This file downsample.c is part of rtl_hpsdr.
 *
 *	rtl_hpsdr - an RTL to HPSDR software translation server
 *	Copyright (C) 2014 Richard Koch
 *
 *	rtl_hpsdr is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	rtl_hpsdr is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with rtl_hpsdr.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "rtl_hpsdr.h"

#ifdef INCLUDE_NEON
#define vector_type float32x4_t
#define vector_zero vmovq_n_f32(0)
#define vector_mac(x,y,z) vmlaq_f32((x), vld1q_f32((y)), (z))
#define vector_store(x,y,z) vst1q_f32((x), vaddq_f32((y), vcombine_f32(vget_low_f32((z)), vget_high_f32((y)))))
#elif defined INCLUDE_SSE2
#define vector_type __m128
#define vector_zero _mm_setzero_ps()
#define vector_mac(x,y,z) _mm_add_ps((x), _mm_mul_ps(_mm_load_ps((y)), (z)))
#define vector_store(x,y,z) _mm_store_ps((x), _mm_add_ps((y), _mm_shuffle_ps((y), (y), _MM_SHUFFLE(1,0,3,2))))
#else
#define vector_type float
#endif

void
down_loop(struct rcvr_cb* rcb, int pass) {
	int i, j, k = 0;
	int dsample, count, coeff_len;
	bool IQ_swap;
	float* buf, *out;
	float* pSrc, *orig_buf;
	const vector_type* pFil;
	vector_type sum1;//, sum2;
	struct main_cb* mcb = rcb->mcb;

	if(1 == pass) {
		// 1st pass filter is hb for 768000
		IQ_swap = true;
		dsample = 2;
		count = RTL_READ_COUNT;
		coeff_len = COEFF3072_H_16_LENGTH;
		buf = &(rcb->iq_buf[0]);
		out = &(rcb->iq_buf_final[COEFF1536_H_32_LENGTH * 2]);
	} else {
		IQ_swap = false;
		count = RTL_READ_COUNT / 2;
		coeff_len = mcb->length_fir;
		buf = &(rcb->iq_buf_final[0]);
		out = &(rcb->iqSamples[rcb->iqSamples_remaining * 2]);

		switch(mcb->output_rate) {
		case 48000:
			dsample = DOWNSAMPLE_192 * 2;
			break;

		case 96000:
			dsample = DOWNSAMPLE_192;
			break;

		case 192000:
			dsample = DOWNSAMPLE_192 / 2;
			break;

		case 384000:
			dsample = DOWNSAMPLE_192 / 4;
			break;
		}
	}

	orig_buf = buf;

#if defined(INCLUDE_NEON) || defined(INCLUDE_SSE2)

	// filter is evaluated for two I/Q samples with each iteration, thus use of 'j += 2'
	for(j = 0; j < count / 2; j += 2) {
		pSrc = buf;

		// filter coefficients. NOTE: Assumes coefficients are aligned to 16-byte boundary
		if(1 == pass)
			pFil = (const vector_type*) mcb->align3072_768_H;
		else {
			switch(mcb->output_rate) {
			case 48000:
				pFil = (const vector_type*) mcb->align1536_48_H;
				break;

			case 96000:
				pFil = (const vector_type*) mcb->align1536_96_H;
				break;

			case 192000:
				pFil = (const vector_type*) mcb->align1536_192_H;
				break;

			case 384000:
				pFil = (const vector_type*) mcb->align1536_384_H;
				break;
			}
		}

		sum1 = vector_zero;
//        sum1 = sum2 = vector_zero;

		for(i = 0; i < coeff_len / 8; i++) {
			// Unroll loop for efficiency & calculate filter for 2*2 I/Q samples
			// at each pass

			// sum1 is accu for 2*2 filtered I/Q data at the primary data offset
			// sum2 is accu for 2*2 filtered I/Q data for the next sample offset.
			sum1 = vector_mac(sum1, pSrc, pFil[0]);
//            sum2 = vector_mac(sum2, pSrc+2, pFil[0]);

			sum1 = vector_mac(sum1, pSrc + 4, pFil[1]);
//            sum2 = vector_mac(sum2, pSrc+6, pFil[1]);

			sum1 = vector_mac(sum1, pSrc + 8, pFil[2]);
//            sum2 = vector_mac(sum2, pSrc+10, pFil[2]);

			sum1 = vector_mac(sum1, pSrc + 12, pFil[3]);
//            sum2 = vector_mac(sum2, pSrc+14, pFil[3]);

			pSrc += 16;
			pFil += 4;
		}

		buf += 4;

		// Now sum1 and sum2 both have a filtered 2-channel sample each, but we still need
		// to sum the two hi- and lo-floats of these registers together. Only perform this
		// if we actually need it for the downsampled output data.
		// UPDATE: since we're throwing away sum2, don't bother calculating it.
		if(0 == ((j + 2) % dsample)) {
			// post-shuffle & add the filtered values and store to dest.
			vector_store(rcb->dest, sum1, sum1);
#if 0
			_mm_store_ps(rcb->dest, _mm_add_ps(_mm_shuffle_ps(sum1, sum2, _MM_SHUFFLE(1, 0, 3, 2)),       // s2_1 s2_0 s1_3 s1_2
			                                   _mm_shuffle_ps(sum1, sum2, _MM_SHUFFLE(3, 2, 1, 0))      // s2_3 s2_2 s1_1 s1_0
			                                  ));
#endif

			// Since we're decimating we only need one of the sums
			out[k++] = (IQ_swap) ? rcb->dest[0] : rcb->dest[1];
			out[k++] = (IQ_swap) ? rcb->dest[1] : rcb->dest[0];

#else // non SIMD

	for(j = 0; j < count; j += 2) {

		pSrc = buf;

		if(1 == pass)
			pFil = mcb->coeff_768;
		else {
			switch(mcb->output_rate) {
			case 48000:
				pFil = mcb->coeff_48;
				break;

			case 96000:
				pFil = mcb->coeff_96;
				break;

			case 192000:
				pFil = mcb->coeff_192;
				break;

			case 384000:
				pFil = mcb->coeff_384;
				break;
			}
		}

		sum1 = sum2 = 0.0f;

		for(i = 0; i < coeff_len; i++) {
			sum1 += pSrc[0] * pFil[0];
			sum2 += pSrc[1] * pFil[0];
			pSrc += 2;
			pFil++;
		}

		buf += 2;

		if(0 == (((j + 2) / 2) % dsample)) {
			out[k++] = (IQ_swap) ? sum2 : sum1;
			out[k++] = (IQ_swap) ? sum1 : sum2;
#endif
		}
	}

	// Move the last coeff_len*2 length of buffer to the front for the next call
	memmove(orig_buf, &orig_buf[count - (coeff_len * 2)],
	        coeff_len * 2 * sizeof(float));
}

void
downsample(struct rcvr_cb * rcb) {
	down_loop(rcb, 1);
	down_loop(rcb, 2);
}


rtl_hpsdr.h
#ifndef _RTL_HPSDR_H
#define _RTL_HPSDR_H

/*	This file rtl_hpsdr.h is part of rtl_hpsdr.
 *
 *	rtl_hpsdr - an RTL to HPSDR software translation server
 *	Copyright (C) 2014 Richard Koch
 *
 *	rtl_hpsdr is free software: you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation, either version 3 of the License, or
 *	(at your option) any later version.
 *
 *	rtl_hpsdr is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License
 *	along with rtl_hpsdr.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <stdlib.h>
#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <sys/timeb.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <net/if_arp.h>
#include <net/if.h>
#include <ifaddrs.h>
#include <pthread.h>
#include <stdbool.h>
#include <semaphore.h>
#include <time.h>
#include <signal.h>
#include <unistd.h>
#include <math.h>
#include <limits.h>
#include <string.h>
#include <libgen.h>
#include <errno.h>
#include <rtl-sdr.h>
#ifdef INCLUDE_NEON
#include <arm_neon.h>
#elif defined INCLUDE_SSE2
#include <xmmintrin.h>
#endif

#define PRG_VERSION "1.3" // see ChangeLog for history

#define HERMES_FW_VER 25    //2.5
#define PORT 1024
#define MAX_BUFFER_LEN 2048
#define HPSDR_FRAME_LEN 1032
#define IQ_FRAME_DATA_LEN 63
#define DOWNSAMPLE_192 8    // downsample value used to get 192khz
#define RTL_SAMPLE_RATE (192000 * DOWNSAMPLE_192)
#define RTL_READ_COUNT (2048 * DOWNSAMPLE_192)
#define MAX_RCVRS 8     // cuSDR64 limits this to 7
#define IQ_FRAME_DATA_LEN 63
#define MAXSTR 64

#define COEFF3072_H_16_LENGTH 16
#define COEFF1536_H_32_LENGTH 32
#define COEFF1536_H_64_LENGTH 64
float coeff3072_768_H_16[COEFF3072_H_16_LENGTH];
float coeff1536_384_H_32[COEFF1536_H_32_LENGTH];
float coeff1536_192_H_32[COEFF1536_H_32_LENGTH];
float coeff1536_96_H_32[COEFF1536_H_32_LENGTH];
float coeff1536_48_H_32[COEFF1536_H_32_LENGTH];
float coeff1536_384_H_64[COEFF1536_H_64_LENGTH];
float coeff1536_192_H_64[COEFF1536_H_64_LENGTH];
float coeff1536_96_H_64[COEFF1536_H_64_LENGTH];
float coeff1536_48_H_64[COEFF1536_H_64_LENGTH];

struct main_cb {
	int total_num_rcvrs;
	int active_num_rcvrs;
	u_int rcvrs_mask;
	int nsamps_packet;
	int frame_offset1;
	int frame_offset2;
	int output_rate;
	int length_fir;
	char sound_dev[32];
	char net_dev[32];

	// the last array member is used to remember last settings
	int agc_mode[MAX_RCVRS + 1];
	int direct_mode[MAX_RCVRS + 1];
	int gain[MAX_RCVRS + 1];
	int freq_offset[MAX_RCVRS + 1];
	int rcvr_order[MAX_RCVRS + 1];
	int signal_multiplier[MAX_RCVRS + 1];

	struct timeb freq_ltime[MAX_RCVRS];
	struct timeb freq_ttime[MAX_RCVRS];

	// filter coefficient table pointers
	float* align1536_48_H;
	float* align1536_96_H;
	float* align1536_192_H;
	float* align1536_384_H;
	float* align3072_768_H;
	float* coeff_48;
	float* coeff_96;
	float* coeff_192;
	float* coeff_384;
	float* coeff_768;

	struct rcvr_cb {
		float dest[4] __attribute__((aligned(16)));

		int rcvr_num;
		int freq;
		u_int rcvr_mask;
		rtlsdr_dev_t* rtldev;
		struct main_cb* mcb;
		pthread_t hpsdrsim_sendiq_thr;
		pthread_t rtl_read_thr;

		int iqSample_offset;
		int iqSamples_remaining;
		float iqSamples[(RTL_READ_COUNT + (IQ_FRAME_DATA_LEN * 2))];

		u_char rtl_buf[RTL_READ_COUNT];
		float* iq_buf;
		float* iq_buf_final;
	} rcb[MAX_RCVRS];
};

void downsample(struct rcvr_cb* rcb);
void format_payload(void);
int init_rtl(int rcvr_num, int dev_index);
void load_packet(struct rcvr_cb* rcb);
void rtl_sighandler(int signum);
int get_addr(int sock, char* ifname);
void hpsdrsim_reveal(void);

pthread_t hpsdrsim_thread_id;
void* hpsdrsim_thread(void* arg);
pthread_t watchdog_thread_id;
void* hpsdrsim_watchdog_thread(void* arg);
void* hpsdrsim_sendiq_thr_func(void* arg);
void* rtl_read_thr_func(void* arg);
void hpsdrsim_stop_threads();

void open_local_sound(char* adevice);
void close_local_sound();
void write_local_sound(unsigned char* samples);
void reopen_local_sound();

#endif // _RTL_HPSDR_H
